### ISSUE TITLE
Sparse matrix multiplication

### ISSUE THREAD

**Issue by abhinav28071999** [sympy/sympy#18662]
sparse matrix multiplication in sympy uses naive multiplication algorithm. If i am right about this then the following algorithm can be implemented for sparse matrices.
https://www.degruyter.com/downloadpdf/j/comp.2014.4.issue-1/s13537-014-0201-x/s13537-014-0201-x.pdf
Let me know about this and i will submit a suitable pr.

**Comment by sylee957** [comment]
I don't think that what was there is that much naive https://github.com/sympy/sympy/blob/cd86e3c3335a7f43379185c239619c576522ef4a/sympy/matrices/sparse.py#L375
So I think that it's worth to put under a different method of `Matrix.multiply` and investigate which bounds would make that faster than the original.

I can only verify whether the multiplication is done right or wrong.

**Comment by Saanidhyavats** [comment]
It will be good if we can make the changes in the same file `sympy/sympy/matrices/sparse.py` with a new method . Both the methods shall be there and later on one can be removed based on the performance. @sylee957  shall I make the changes in the same file?

**Comment by sylee957** [comment]
Yes, you should edit `SparseMatrix.multiply` and first implement them as different methods, and see if we can bind each ones to the matrix core.

**Comment by Saanidhyavats** [comment]
I am working on this if no one is currently working on it.

**Comment by ks147** [comment]
@sylee957 is this issue closed? Is there a new PR on this yet?

**Comment by oscarbenjamin** [comment]
The OP link for the paper doesn't work but I think it's this:
https://www.degruyter.com/document/doi/10.2478/s13537-014-0201-x/html

It's not clear to me that the current implementation is substantially different from what is described in the linked paper. Note that the storage model for sparse matrices in sympy is not the same as described in the paper.

There is also a sparse implementation of DomainMatrix now which uses essentially the same algorithm as SparseMatrix:
https://github.com/sympy/sympy/blob/a96ca96e6505b71a846f266884c09197f71e7232/sympy/polys/matrices/sdm.py#L97-L116
